puresnmp/__init__.py: py2

remdragon opened the pull request:
I have implemented py2 compatibility while making sure that all tests still work in py3

I noticed while going through it that you wanted NonASN1Type renamed to UnknownType so I took care of that, too.

While I was troubleshooting my changes, I had a desire to see the ascii representation of each byte in ByteTester.assertBytesEqual so I added an extra column with that information.

puresnmp/exc.py: py2
puresnmp/pdu.py: py2
puresnmp/test/__init__.py: py2, add repr() of each char to comparison output
puresnmp/tests/test_non_standard_types.py: py2
puresnmp/test/test_package_root.py: py2, switch some base classes to ByteTester b/c I implemented py2 unittest compatibility there
puresnmp/test/test_pdus.py: py2
puresnmp/test/x690/test_types.py: py2, rename NonASN1Type to UnknownType
puresnmp/test/x690/test_util.py: py2
puresnmp/transport.py: py2
puresnmp/types.py: py2
puresnmp/x690/types.py: py2
puresnmp/x690/util.py: py2

exhuma commented:
Well this puts me in a dilemma... I've planned to implement asyncio in the future which will definitely not be Python 2 compatible.

*However*, I am currently working on other projects which have priority so the asyncio support will definitely come later rather than sooner.

So in the mean-time I am OK with merging this. But I may have to revert this at some point in the future. I'll go over the PR in a moment.

exhuma commented on puresnmp/test/__init__.py:
```diff
@@ -74,4 +95,4 @@ def readbytes(filename):
         str_bytes.extend(line.split())
     values = [int(char, 16) for char in str_bytes]
-    return bytes(values)
+    return bytearray(values)
```
Why did you replace `bytes` with `bytearray` here? `bytes` also exists in Python2.

exhuma replied:
... or to be more precise, there is a tiny difference in `bytes` and `bytearray`. Was there a reason you needed a `bytearray` instead of `bytes`? I don't think it does make a big difference here so I am fine either way. I'm just curious.

exhuma commented on puresnmp/x690/types.py:
```diff
@@ -93,7 +104,8 @@ def pop_tlv(data):
 
 
-class Type(metaclass=Registry):
+@six.add_metaclass ( Registry )
```
Can you run a PEP-8 checker on your code? I've seen these spurious spaces around identifiers and parens on various places which should be removed to have a more consistent code-style

remdragon replied:
There are things about PEP-8 that really stick in my craw, so it's not something I following closely in my personal projects. Unfortunately it means I'm not very good at it when submitting patches to other projects. I will look into how to run a PEP-8 check and do that.

exhuma replied:
I get that. It is difficult when a tool imposes a code-style onto you that you don't like. I have my issues with certain things in PEP8 as well. Then again, PEP8 itself statees that it is only a guideline and that you can break it if you have a good reason. After all, it's about readability. And sometimes PEP8 *does* make the code less readable.

Having said that, the issue I have with the spaces around the functions is primarily that it is inconsistent with the rest of the code in this project 😉 

One advantage here of PEP8 is that there are many checkers available which you can easily run.

remdragon replied:
yes and I fixed up most of the spaces that I had introduced, but I just apparently missed some. As I said I had not intended to break the coding style of the project in any way, I just wasn't careful enough. I'm working on it now.

exhuma commented on puresnmp/x690/util.py:
```diff
@@ -48,8 +63,8 @@ def from_bytes(data: Union[int, bytes]) -> "TypeInfo":
 
-        if isinstance(data, bytes):
-            data = int.from_bytes(data, 'big')
+        if isinstance(data, ( bytes, bytearray)):
```
To add to my previous comment about `bytes` and `bytearrays`: I generally try to pay attention to mutability. Introducing `bytearray` makes the value mutable. Essentially using `bytearray` changes the data-type to something which allows dangerous operations. Using `str` instead of `bytearray` would keep the same semantics & behaviour. Unfortunately `str` has different meaning whether you look at it from Python 2 or Python 3. After giving it some thought, I would suggest using [six.binary_type](https://pythonhosted.org/six/#six.binary_type) instead of `bytearray`.

On a bytes-level getting Python2/3 compatibility can be tricky.

remdragon replied:
I'm going to respond here in general to your multiple questions about bytes. I ran into 2 problems with bytes when converting the code to Python 2:

1) bytes(int) behaves very difficult in py2 than py3. However, bytearray(int) behaves the same way, so it seemed the simplest fix.

2) I like your suggestion of overriding bytes() better than using bytearray, so I will get to work making that change.

remdragon replied:
I just googled it, and six.binary_type unfortunately won't fix the problem because of bytes(int) behavior, but creating a bytes function that behaves as desired should do the trick.

remdragon replied:
I'm currently working through the code with your suggestion of overriding bytes, and I will follow it through to completion, but a couple things are bothering me about this approach:

1) Overriding a native type could have unintended consequences and be difficult for other programmers to understand why we chose to do it that way.

2) bytearray has more consistent behavior between py2 and py3

3) bytearray actually performs much better than bytes, see here for example:

http://ze.phyr.us/bytearray/

remdragon replied:
Do you know how to fix it where I can override bytes in py2 without having to rewrite any isinstance ( foo, bytes ) calls?

At the moment it's looking like I'd have to add a bytes_type = bytes and then use bytes_type in every isinstance call.

You wanted less noise, I'm feeling more and more like this approach will not result in less noise, but maybe I'm just not a good enough Python programmer yet :(

remdragon replied:
I'm looking more closely at "six" and I see a few functions that will make my diff cleaner. In particular six.byte2int() and six.indexbytes() will eliminate several uses of bytearray()

remdragon replied:
Okay, I think I found a solution, would you glance at this and give me your feedback?

```python
import six
import sys

def repr2(s):
	# make sure strings always have type prefix ( for the purposes of validating this logic )
	if not isinstance(s, str):
		return repr(s)
	if six.PY2:
		return 'b'+repr(s)
	else:
		return 'u'+repr(s)

try:
	unicode
except NameError:
	unicode = str

# modifications to puresnmp.x690.types.Type class
# so that all classes will inherit the needed behavior
# note that repr2 is only for validation here, we'd use the real repr() in puresnmp
class Type(object):
	def __repr__(self):
		return u'Type({})'.format(repr2(b'foo'))
	def __bytes__(self):
		return b'foo'
	if six.PY2:
		def __unicode__(self):
			return self.__repr__()
		def __str__(self):
			return self.__bytes__()

obj = Type()

def assertEquals ( a, b ):
	assert a == b, ('Assertion Error:\n'
		'   {!r}\n'
		'!= {!r}'.format ( a, b )
	)
assertEquals ( repr2(unicode(obj)), '''u"Type(b'foo')"''' )
b = bytes(obj)
assertEquals ( repr2(b), "b'foo'" )
assertEquals ( repr2(six.byte2int(b)), '102' )
assertEquals ( repr2(six.indexbytes(b,1)), '111' )
```

remdragon replied:
The only complication left is that bytes([40]) does something different on py2.

I feel very strongly against overriding the built-in bytes, so here's what I propose:

1) change those lines from bytes(foo) to to_bytes(foo):

```python
def to_bytes(x):
    return bytes(bytearray(x))
```

2) In order to avoid adding an isinstance() check to to_bytes(), we would need to add one more method to puresnmp.x690.types.Type:

```python
	def __iter__(self):
		return six.iterbytes(self.__bytes__())
```

Would the proposed modifications be acceptable to you?

exhuma commented on puresnmp/x690/util.py:
```diff
@@ -75,7 +90,7 @@ def from_bytes(data: Union[int, bytes]) -> "TypeInfo":
         return instance
 
-    def __bytes__(self):
+    def to_bytes(self):
```
Considering that at some point in the future (once asyncio hits) the library will no longer work in Python 2 I think it would be prudent to cut down on the size of the diff. One way to do this is revert the replacement of `__bytes__` with `to_bytes` and introduce a new `bytes()` function for Python2. In a similar way you did with the `int_from_bytes` function. This function could simply delegate the call to `__bytes__` of the object. This way the code would not need to be modified on so many places.

Something along the lines of:

    def bytes(obj):
        return obj.__bytes__()

remdragon replied:
out of all the snmp libraries I reviewed, yours was the only one that ticked all of my boxes, except the fact that it didn't support py2 (yet). I was having a real problem with inconsistent behavior w/r/t MIB files, problems supporting the Windows environment, etc.

The problem with py3 for me is that I have a need to drop executable binaries on client servers that are being rented and the vendor they are renting from do not allow them to run installers for fear of introducing incompatibilities with the vendor's software.

Building python py3 executables with pyinstaller on windows 10 currently requires installation of the Windows UCRT on the destination machine.

Also, I use wxPython extensively and only the phoenix version is supported on Python 3, but phoenix is still very unstable.

For these reasons I am unfortunately, sadly, and painfully stuck with py2, but it pays the bills if you know what I mean.

I haven't looked into asyncio much but when I have, I've had a hard time understanding why people love it so much. Would you mind sharing with me why you want to introduce asyncio ( I'd love to better understand it's appeal ), and can I help you come up with a way of doing it that doesn't break py2 compatibility? I could always fork your project for what I'm usually your library for, but I'd rather be able to take advantage of future bug fixes.

exhuma replied:
> I was having a real problem with inconsistent behavior w/r/t MIB files, problems supporting the Windows environment, etc.

Exactly why I started this project

> Would you mind sharing with me why you want to introduce asyncio 

Well, for starters this project was my sandbox for learning `asyncio`. I wanted to get it working first, and then introduce `asyncio`. I have not gotten around to that yet due to other projects having priority. But one day I will. But there is a chance that it can be done in a compatible manner. I will have to see. The biggest problem is that `asyncio` introduced the new `async`/`away` syntax which does not exist in Python 2. So it will break Python 2 in a fundamental manner. Maybe there are some ways around it. I'll have to see.

But it may take some time until I get around to that, so you are still good 😉 and a fork is not (yet) necessary, and is also something I would like to avoid if possible.

A simple way to explain the appeal to `asyncio` is that it allows you to easily tell the runtime: "Listen, I'm going to do I/O now, so you can do something else while I wait for a response". In essence it is *similar* (but definitely not the same) as parallelisation (threading & co), but it is easier as you don't have to deal with locks. One way of thinking about it is: Using parallelisation, context-switches are imposed by the kernel and you never know when they happen. Using `asyncio` *you* decide when to switch.

It is a really interesting model which is *perfect* for a network library like `puresnmp`. And many applications using it could really benefit from this. Alas, it is Python3 only.

remdragon replied:
I wonder if there's a way we could introduce an opt-in asyncio wrapper in puresnmp that py3 clients can use while not introducing the breaking syntax in the core snmp logic, kinda like how you explicitly tell people that MIB processing is not within scope of the project?

remdragon commented:
I couldn't bring myself to redefine bytes(), it just seemed like there could be too many unintended consequences and create confusion further down the road, but I was able to eliminate bytearray in production code paths except for py2's to_bytes() implementation itself. However, I immediately convert it to a bytes object before returning it.

I'm mostly confident I fixed all the PEP8 violations I introduced. However, there are a lot in your code already, and I didn't want to create extra noise in my diff by fixing them all for you. It wouldn't take me long to do it for you if you like, but I think that would best be done as a separate patch for code history reasons.

exhuma commented:
> I couldn't bring myself to redefine bytes(), it just seemed like there could be too many unintended consequences ( like breaking isinstance checks in particular ) and create confusion further down the road, but I was able to eliminate bytearray in production code paths except for py2's to_bytes() implementation itself. However, I immediately convert it to a bytes object before returning it.

I agree. Sounds reasonable

> I'm mostly confident I fixed all the PEP8 violations I introduced. However, there are a lot in your code already, and I didn't want to create extra noise in my diff by fixing them all for you. It wouldn't take me long to do it for you if you like, but I think that would best be done as a separate patch for code history reasons.

I agree to not include them here. It would be too much noise in the commit log. Additionally, you would show up as the last editor of that line even though you did not make a "logical" change. So I consider these to be on me. I quickly looked over my code but could not find *that* many important violations. Mostly they are about line-length. Maybe I had my editor set up to 80 instead 79 when writing this. But as I mentioned earlier, the goal is readability and consistency. Blindly following each recommendation of the linters is not *always* helping readability.

One last thing though: When I merge this, I will have to release it as a major release. I am following semantic versioning, and the rename from `NonASN1Type` to `UnknownType` changes the external API which will force me to bump the major. I wanted to leave such a small change for a later release so I could include a bit more in a major version bump. Also, I am not 100% sure of all the potential implications of the byte/bytearray change, so I think a major bump is anyway a prudent choice.

I have to head out now, but I will merge this in the coming days. If I fall silent, feel free to nudge me.

remdragon commented:
On Sun, Feb 11, 2018 at 12:02 PM Michel Albert <notifications@github.com>
wrote:

> I couldn't bring myself to redefine bytes(), it just seemed like there
> could be too many unintended consequences ( like breaking isinstance checks
> in particular ) and create confusion further down the road, but I was able
> to eliminate bytearray in production code paths except for py2's to_bytes()
> implementation itself. However, I immediately convert it to a bytes object
> before returning it.
>
> I agree. Sounds reasonable
>
> I'm mostly confident I fixed all the PEP8 violations I introduced.
> However, there are a lot in your code already, and I didn't want to create
> extra noise in my diff by fixing them all for you. It wouldn't take me long
> to do it for you if you like, but I think that would best be done as a
> separate patch for code history reasons.
>
> I agree to not include them here. It would be too much noise in the commit
> log. Additionally, you would show up as the last editor of that line even
> though you did not make a "logical" change. So I consider these to be on
> me. I quickly looked over my code but could not find *that* many
> important violations. Mostly they are about line-length. Maybe I had my
> editor set up to 80 instead 79 when writing this. But as I mentioned
> earlier, the goal is readability and consistency. Blindly following each
> recommendation of the linters is not *always* helping readability.
>

It was rather arduous trying to pick through the pep8 analysis logs and
make sure I identified all of my changes. I humbly request you consider
fixing those issues, even if they’re not that important, if you’re going to
ask other people to conform to pep8 when submitting changes.

> One last thing though: When I merge this, I will have to release it as a
> major release. I am following semantic versioning, and the rename from
> NonASN1Type to UnknownType changes the external API which will force me
> to bump the major. I wanted to leave such a small change for a later
> release so I could include a bit more in a major version bump. Also, I am
> not 100% sure of all the potential implications of the byte/bytearray
> change, so I think a major bump is anyway a prudent choice.
>

I’m not disagreeing, just providing some food for thought:

NonASN1Type is still in there as an alias to UnknownType, so it shouldn’t
break any existing code unless they are depending on the output of repr()
for those objects.

Actually, I did the alias wrong. I need to create it as a subclass of
UnknownType so that repr() will output the correct value. Also, I think the
subclass will be necessary in order to flag it as deprecated anyway. Please
give me a chance to fix that.

I think I eliminated bytearray from everything except py2’s to_bytes() and
the unit test code which didn’t seem that important to fix. If I missed
one, I will look at fixing it, as I didn’t mean to leave any in there.

Are there security implications in immediately converting a bytearray back
to a bytes?

Why is a bytearray worse than your current approach of a list of int? If we
switched to using bytearray in place of where you are currently using lists
of int, we might be able to eliminate to_bytes(). We could add unittests to
make sure that bytearrays never come out of the public api if that’s the
concern. Forgive me if I seem like I’m beating a dead horse, I’m just
trying to understand and google isn’t helping much.

I have to head out now, but I will merge this in the coming days. If I fall
> silent, feel free to nudge me.
>
Are you going to add 2.7 to your continuous integration tests once you do
merge it?

—
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/exhuma/puresnmp/pull/32#issuecomment-364772551>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADka11YvNBYvgSEQBiNROG6DnhH4TOb9ks5tTysmgaJpZM4SATbk>
> .
>
-- 


Royce Mitchell, IT Consultant
ITAS Solutions
royce3@itas-solutions.com

exhuma commented:
> [...] I humbly request you consider fixing those issues, even if they’re not that important, if you’re going to ask other people to conform to pep8 when submitting changes.

Absolutely. I already feel like an idiot for asking for PEP8 compliance while my own code does not conform to it 100%. However, I do *not* see PEP8 as a very strict rule. Unfortunately every tool has it's own markup to disable checks on certain lines. I think I should add a section to the contribution guide about which tool I use in this project to avoid confusion/annoyance for future contributors.

> NonASN1Type is still in there as an alias to UnknownType, so it shouldn’t break any existing code unless they are depending on the output of repr() for those objects.

Ah cool. Did not spot that 👍 

> [...] Please give me a chance to fix that.

Sure. Just give me a nudge when you a ready for merge. This is also why I did not merge prematurely yesterday. I'll leave some time for the PR to settle.

> Are there security implications in immediately converting a bytearray back to a bytes?

Not a *security* hole as far as I am aware. With "dangerous" I meant the fact that it would change an immutable type to a mutable type. And you are correct: if this does not leak to the outside of the API it will be perfectly fine to change. I have not looked at the code in quite some time so some details escape me. I'll leave that choice to your discretion then, as you have the code actively in your head and may spot things I don't at the moment. If it makes things easier, by all means, change it!

exhuma commented:
> Are you going to add 2.7 to your continuous integration tests once you do merge it?

Yes 😉

remdragon commented:
Okay, so there's definitely more I could do in this PR, but software is never done, and this PR successfully adapts puresnmp to be py2 compatible. If you're satisfied, let's call this done and other improvements can be left for separate PR's.

exhuma commented:
Just wanted to give a sign of life of myself. I'm still willing to work on this but as already mentioned I need to finish other projects first. But rest assured that I will look after this.

remdragon commented:
Thank you!

On Sun, Mar 18, 2018 at 1:35 PM Michel Albert <notifications@github.com>
wrote:

> Just wanted to give a sign of life of myself. I'm still willing to work on
> this but as already mentioned I need to finish other projects first. But
> rest assured that I will look after this.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/exhuma/puresnmp/pull/32#issuecomment-374029722>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADka1_z0U-CbjiQnF7-s4SfzRVBLG4Qcks5tfqjdgaJpZM4SATbk>
> .
>
-- 


Royce Mitchell, IT Consultant
ITAS Solutions
royce3@itas-solutions.com

exhuma commented:
I've come back to life! :wink: 

I'll dive back into the project again in the coming days.

exhuma commented:
@remdragon  Can you have a look at #38 to make sure it's backwards compatible? I'm trying to figure out if it's necessary to bump the major version or not.

In the mean-time I'll sift through my own PEP8 violations and other existing issues.

remdragon commented:
I'm not sure what you're asking. I changed my PR so that NonASN1Type is a
subclass of UnknownType. I believe this makes it 100% backwards compatible.



Royce Mitchell, IT Consultant
ITAS Solutions
royce3@itas-solutions.com

On Fri, May 25, 2018 at 1:58 AM, Michel Albert <notifications@github.com>
wrote:

> @remdragon <https://github.com/remdragon> Can you have a look at #38
> <https://github.com/exhuma/puresnmp/issues/38> to make sure it's
> backwards compatible? I'm trying to figure out if it's necessary to bump
> the major version or not.
>
> In the mean-time I'll sift through my own PEP8 violations and other
> existing issues.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/exhuma/puresnmp/pull/32#issuecomment-391960416>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADka16IHI86GY_Hps1bONx1vfi-nSS7fks5t16t_gaJpZM4SATbk>
> .
>

exhuma commented:
Sorry. I've misread [your older comment](https://github.com/exhuma/puresnmp/pull/32#issuecomment-365930773), I somehow read it as if you wanted to do more in this PR.

I'll merge.

exhuma commented:
I am currently in the process of preparing a new release with your changes.

However, I also started actively using `puresnmp` as a replacement for an existing SNMP tool at the office and have run into several difficulties introduced by the PY2 compatibility.

Given that Python 2 is EOL in 2020, I am considering rolling back the PY2 changes in the near future and publishing it as `puresnmp2`. This will ensure a much cleaner code-base for the future.

@remdragon Are you still restricted by Python2? Or did you have the chance to upgrade to 3 since this PR was created?

remdragon commented:
I am still restricted to py2. A recent re-evaluation yields that
pyinstaller still can't generate a drop-in package from windows 10 that
will run on older versions of Windows without having to run an extra
installer.

Also, last time I checked, the http server in py3 is still broken, and I
just don't have the time to figure out a fix and submit it to them.

I still can't use puresnmp as published because of the pyinstaller issue
with version.txt and I strongly dislike having extra unneeded files in my
distributables. Since you don't want to fix that, it looks like I will be
forced to continue using my fork of puresnmp, which is sad because it will
make it difficult to share bug fixes going forward.

I can't promise I have a lot of time to help, but what py3 issues are you
experiencing? I can try running my code in py3 and see what issues I
experience.



Royce Mitchell, IT Consultant
ITAS Solutions
royce3@itas-solutions.com

On Thu, Jun 7, 2018 at 1:33 AM, Michel Albert <notifications@github.com>
wrote:

> I am currently in the process of preparing a new release with your changes.
>
> However, I also started actively using puresnmp as a replacement for an
> existing SNMP tool at the office and have run into several difficulties
> introduced by the PY2 compatibility.
>
> Given that Python 2 is EOL in 2020, I am considering rolling back the PY2
> changes in the near future and publishing it as puresnmp2. This will
> ensure a much cleaner code-base for the future.
>
> @remdragon <https://github.com/remdragon> Are you still restricted by
> Python2? Or did you have the chance to upgrade to 3 since this PR was
> created?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/exhuma/puresnmp/pull/32#issuecomment-395308846>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADka11ED6CAGLJcCO7lw-tzeSW4FcgX0ks5t6MksgaJpZM4SATbk>
> .
>

remdragon commented:
also, I've recently being doing a lot with mypy. A quick glance shows that
there are multiple mypy syntax errors in my copy of the puresnmp code base.



Royce Mitchell, IT Consultant
ITAS Solutions
royce3@itas-solutions.com

On Thu, Jun 7, 2018 at 6:13 AM, Royce Mitchell III <royce3@gmail.com> wrote:

> I am still restricted to py2. A recent re-evaluation yields that
> pyinstaller still can't generate a drop-in package from windows 10 that
> will run on older versions of Windows without having to run an extra
> installer.
>
> Also, last time I checked, the http server in py3 is still broken, and I
> just don't have the time to figure out a fix and submit it to them.
>
> I still can't use puresnmp as published because of the pyinstaller issue
> with version.txt and I strongly dislike having extra unneeded files in my
> distributables. Since you don't want to fix that, it looks like I will be
> forced to continue using my fork of puresnmp, which is sad because it will
> make it difficult to share bug fixes going forward.
>
> I can't promise I have a lot of time to help, but what py3 issues are you
> experiencing? I can try running my code in py3 and see what issues I
> experience.
>
>
>
> Royce Mitchell, IT Consultant
> ITAS Solutions
> royce3@itas-solutions.com
>
> On Thu, Jun 7, 2018 at 1:33 AM, Michel Albert <notifications@github.com>
> wrote:
>
>> I am currently in the process of preparing a new release with your
>> changes.
>>
>> However, I also started actively using puresnmp as a replacement for an
>> existing SNMP tool at the office and have run into several difficulties
>> introduced by the PY2 compatibility.
>>
>> Given that Python 2 is EOL in 2020, I am considering rolling back the PY2
>> changes in the near future and publishing it as puresnmp2. This will
>> ensure a much cleaner code-base for the future.
>>
>> @remdragon <https://github.com/remdragon> Are you still restricted by
>> Python2? Or did you have the chance to upgrade to 3 since this PR was
>> created?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/exhuma/puresnmp/pull/32#issuecomment-395308846>, or mute
>> the thread
>> <https://github.com/notifications/unsubscribe-auth/ADka11ED6CAGLJcCO7lw-tzeSW4FcgX0ks5t6MksgaJpZM4SATbk>
>> .
>>
>
>

exhuma commented:
I agree with your comments. Don't get me wrong. I actually want to make it work for you. Let me addres s the points you brought up in a second... first, I want to mention that I have pushed a release candidate with your Python 2 changes to pypi. I did not have time to test it in a Python 2 environment yet (hence being an RC release): https://pypi.org/project/puresnmp/#history (https://github.com/exhuma/puresnmp/tree/release/1.3.0)

On to your points:

## pyinstaller 
I unfortunately don't have a Windows development environment handy to test this out so I have to rely on information I get from elsewhere. From what you say the fix brought up in the issues does not yet solve the problem. I will revisit the issue and see what I can do.

## version.txt
I am also not adamantly fixed on having a `version.txt` in the repo. But my past experience has shown that it is very beneficial to have the version information available in a file which is easily pareseable (especially for deployment cycles and other devops work).

I think this is a matter of personal preference and habit. But I'm always open for new ideas. How do you extract the version info from source code with external programs? Initially I used a `cat`/`grep`/`sed` solution extracting the info from `__init__.py` but found that having a separate file makes this task easier. I also looked into `git describe` but was not all too happy with the output, but it might be a viable alternative.

## Python 2/3
There are no breaking issues. It is mainly related to code-quality and readability. And given that Python 2 is going EOL soon I'm a bit reluctant on introducing additional code for that. But again, this is not a deal-breaker. If you still need Python 2, I will leave it in.

Just like the `version.txt` issue, I am not hell-bent on my comments above. I believe we both have good intentions but just simply different work-flows & habits. And the project will only benefit from having a common code-base instead of two forks, so I am willing to make compromises ;)

## type-checking
I only recently started using type-checking and am still learning. I saw some remaining issues as well and am aware that some things need to be fixed. Either by using better annotations or refactoring a few things. But I'll keep an eye on that.

exhuma commented:
I've added Python2 to Travis-CI and found some issues. I'll work on those and should have fixed them soon

remdragon commented:
Royce Mitchell, IT Consultant
ITAS Solutions
royce3@itas-solutions.com

On Thu, Jun 7, 2018 at 7:03 AM, Michel Albert <notifications@github.com>
wrote:

> I agree with your comments. Don't get me wrong. I actually want to make it
> work for you. Let me addres s the points you brought up in a second...
> first, I want to mention that I have pushed a release candidate with your
> Python 2 changes to pypi. I did not have time to test it in a Python 2
> environment yet (hence being an RC release): https://pypi.org/project/
> puresnmp/#history (https://github.com/exhuma/puresnmp/tree/release/1.3.0)
>
> On to your points:
> pyinstaller
>
> I unfortunately don't have a Windows development environment handy to test
> this out so I have to rely on information I get from elsewhere. From what
> you say the fix brought up in the issues does not yet solve the problem. I
> will revisit the issue and see what I can do.
>
​Microsoft is certainly pissing me off recently, so much that I'm seriously
considering using Linux more than I do now. Unfortunately my client base is
heavily Windows-centric and the idea of trying to build Windows executables
from a Linux environment does not sound like an improvement in my work-flow.

I've actually seriously considered trying to get python 3 to build with
mingw to resolve the ucrt redistribution issue. I may have to get this
working before the 2020 cut-off. I'd certainly strongly prefer to use
python 3 if I could. I realize this will cause a problem with pip installs
that require a C compiler, but I don't depend on too many of those, so
hopefully I can find alternatives.

> version.txt
>
> I am also not adamantly fixed on having a version.txt in the repo. But my
> past experience has shown that it is very beneficial to have the version
> information available in a file which is easily pareseable (especially for
> deployment cycles and other devops work).
>
> I think this is a matter of personal preference and habit. But I'm always
> open for new ideas. How do you extract the version info from source code
> with external programs? Initially I used a cat/grep/sed solution
> extracting the info from __init__.py but found that having a separate
> file makes this task easier. I also looked into git describe but was not
> all too happy with the output, but it might be a viable alternative.
>
​How about a little 3-line python script?​

    ​from __future__ import print_function
    import puresnmp
    print ( puresnmp.version )

or if you don't want a separate file, add the ability for puresnmp to print
it's version:

    python -m puresnmp --version

This is actually how I'm dealing with packaging nowadays. Importing the
code I'm about to package has the side-benefit that if there's a missing
import, the build will fail. Otherwise, PyInstaller will all too happily
generate an executable that won't actually work.

I am automatically embedding the svn revision level in my executables so
that when I get a bug report, I know exactly which version of the code I'm
having a problem with. This is a messy process involving opening up the
sqlite file in the .svn folder, so I get why you wouldn't want to depend on
something like git describe. With CVS you could have it automatically
update your files with a build id. I don't know why svn/git didn't
implement this, but it's annoying.

Python 2/3
>
> There are no breaking issues. It is mainly related to code-quality and
> readability. And given that Python 2 is going EOL soon I'm a bit reluctant
> on introducing additional code for that. But again, this is not a
> deal-breaker. If you still need Python 2, I will leave it in.
>
Perhaps I don't understand code-quality. I thought it was code that works
in the widest variety of circumstances? I'm always open to learning
something new.

​Readability is of course a tricky endeavor. ​Could it be better served
here with a helper class that abstracts the byte buffer manipulation in a
way that makes it more readable?

Just like the version.txt issue, I am not hell-bent on my comments above. I
> believe we both have good intentions but just simply different work-flows &
> habits. And the project will only benefit from having a common code-base
> instead of two forks, so I am willing to make compromises ;)
> type-checking
>
> I only recently started using type-checking and am still learning. I saw
> some remaining issues as well and am aware that some things need to be
> fixed. Either by using better annotations or refactoring a few things. But
> I'll keep an eye on that.
>
​I have been drinking the mypy kool-aide​

​for the last month. It is not a perfect tool, but it's worth it. I have
caught a number of obscure bugs throughtout my code base. Use --strict and
stay away from Any whenever possible. A quick glance says we want to define
a Union of all the types that can possibly come out of the subclass
overrides of Type.pythonize()​.

> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/exhuma/puresnmp/pull/32#issuecomment-395395998>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADka1wa_scRZNqEUL3LXqZ-aqKEmvYwmks5t6RZ5gaJpZM4SATbk>
> .
>
